src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs

```diff
@@ -95,6 +95,99 @@ public static Uri GenerateSelfLink(this ResourceContext resourceContext, bool in
+                throw Error.Argument("resourceContext", SRResources.UrlHelperNull, typeof(ResourceContext).Name);
+            }
+
+            IList<ODataPathSegment> navigationPathSegments = GenerateODataPathSegmentsForContainedEntity(resourceContext, includeCast);
+
+            if (includeCast)
+            {
+                navigationPathSegments.Add(new TypeSegment(resourceContext.StructuredType, navigationSource: null));
+            }
+
+            navigationPathSegments.Add(new NavigationPropertySegment(navigationProperty, navigationSource: null));
+
+            string link = resourceContext.InternalUrlHelper.CreateODataLink(navigationPathSegments);
+            if (link == null)
+            {
+                return null;
+            }
+
+            return new Uri(link);
+        }
+
+        private static IList<ODataPathSegment> GenerateODataPathSegmentsForContainedEntity(
+            ResourceContext resourceContext, bool includeCast)
+        {
+            IList<ODataPathSegment> odataPathSegments = new List<ODataPathSegment>();
+            Stack<ODataPathSegment> reversedOdataPathSegments = new Stack<ODataPathSegment>();
+
+            ResourceContext currentLevelResourceContext = resourceContext;
+
+            // Loop through the all ExpandedResource(s).
+            // We generate OData Path Segments for all levels.
+            while (currentLevelResourceContext != null)
+            {
+                IEdmNavigationSource navigationSource = currentLevelResourceContext.NavigationSource;
+
+                // For Contained entity sets
+                if (navigationSource is IEdmContainedEntitySet containedEntitySet)
+                {
+                    NavigationPropertySegment navigationPropertySegment = new NavigationPropertySegment(containedEntitySet.NavigationProperty, containedEntitySet.ParentNavigationSource);
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    reversedOdataPathSegments.Push(navigationPropertySegment);
+                }
+
+                // For Non-Contained entity sets
+                if(navigationSource is IEdmEntitySet entitySet)
+                {
+                    EntitySetSegment entitySetSegment = new EntitySetSegment(entitySet);
+                    reversedOdataPathSegments.Push(new KeySegment(ConventionsHelpers.GetEntityKey(currentLevelResourceContext), currentLevelResourceContext.StructuredType as IEdmEntityType, null));
+                    reversedOdataPathSegments.Push(entitySetSegment);
+                }
+
+                // For Singletons
+                if(navigationSource is IEdmSingleton singleton)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    reversedOdataPathSegments.Push(singletonSegment);
+                }
```

Comment:
if we encounter a navigation source that is a singleton, or non-contained entity set (and we have a key), shouldn't we reset the reversedOdatapathSegments in order to return the canonical navigation path?  i.e., if we have an indirect path like users(user1)/groups(group1)/members(user5)/address, where address is a contained navigation property and members is a navigation path to a user (bound to a non-contained entity set "users") then shouldn't we ignore the users(user1)/groups(group1) and just return users(user5)/address?  And similarly, if company is a top level singleton, shouldn't users(user1)/bestFriend/friends(user7)/company/address just return company/address?

What happens if we have groups(group1)/owner/address (where "owner is a single-value non-contained nav prop on a group), would we have to return the (non-canonical) groups(group1)/owner/address, or would we have the id of the owner so that we could return users(userId)/address?